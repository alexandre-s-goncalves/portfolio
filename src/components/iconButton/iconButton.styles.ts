import styled from 'styled-components';

export const ButtonStyled = styled.button`
  display: flex;
  align-items: center;
  justify-content: center;
  border: none;
  outline: none;
  cursor: pointer;
  transition: background-color 0.3s ease;

  &:hover:not(:disabled) {
    opacity: 0.8;
  }

  &:active:not(:disabled) {
    transform: scale(0.95);
  }

  &:disabled {
    cursor: not-allowed;
    opacity: 0.5;
  }

  &:focus-visible {
    outline: 2px solid rgba(0, 0, 0, 0.3);
    outline-offset: 2px;
  }
`;
