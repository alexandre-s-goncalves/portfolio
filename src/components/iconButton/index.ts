export * from './iconButton';
export * from './iconButton.styles';
