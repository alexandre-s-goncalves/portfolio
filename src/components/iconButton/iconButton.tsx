import React from 'react';
import {Icon, IconProps} from 'components/icon';
import {ButtonStyled} from './iconButton.styles';

export interface IconButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  icon: React.ElementType;
  iconColor?: string;
  iconRotation?: number;
  iconWidth?: number;
  iconHeight?: number;
  testId?: string;
}

export const IconButton = ({
  icon,
  iconColor,
  iconRotation = 0,
  iconWidth = 24,
  iconHeight = 24,
  testId,
  ...props
}: Readonly<IconButtonProps>) => {
  const iconProps: IconProps = {
    icon,
    color: iconColor,
    rotation: iconRotation,
    width: iconWidth,
    height: iconHeight,
    cursor: 'inherit',
  };

  return (
    <ButtonStyled data-testid={testId || 'icon-button'} {...props}>
      <Icon {...iconProps} />
    </ButtonStyled>
  );
};
