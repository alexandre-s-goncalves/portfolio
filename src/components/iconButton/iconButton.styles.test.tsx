import React from 'react';
import {render} from '@testing-library/react';
import '@testing-library/jest-dom';
import 'jest-styled-components';
import * as S from './iconButton.styles';

describe('iconButton.styles', () => {
  test('ButtonStyled should match the snapshot', () => {
    const {container} = render(<S.ButtonStyled />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('ButtonStyled renders as a button element', () => {
    const {container} = render(<S.ButtonStyled />);
    const element = container.firstChild as HTMLElement;
    expect(element.nodeName).toBe('BUTTON');
  });

  test('ButtonStyled has cursor pointer by default', () => {
    const {getByTestId} = render(<S.ButtonStyled data-testid="btn" />);
    const button = getByTestId('btn');
    expect(button).toHaveStyle('cursor: pointer');
  });

  test('ButtonStyled with disabled state', () => {
    const {getByTestId} = render(<S.ButtonStyled data-testid="btn" disabled />);
    const button = getByTestId('btn');
    expect(button).toBeDisabled();
  });

  test('ButtonStyled has flex display and center alignment', () => {
    const {getByTestId} = render(<S.ButtonStyled data-testid="btn" />);
    const button = getByTestId('btn');
    expect(button).toHaveStyle('display: flex');
    expect(button).toHaveStyle('align-items: center');
    expect(button).toHaveStyle('justify-content: center');
  });

  test('ButtonStyled has no border and outline', () => {
    const {getByTestId} = render(<S.ButtonStyled data-testid="btn" />);
    const button = getByTestId('btn');
    expect(button).toHaveStyle('border: none');
    expect(button).toHaveStyle('outline: none');
  });

  test('ButtonStyled accepts custom inline styles', () => {
    const {getByTestId} = render(
      <S.ButtonStyled
        data-testid="btn"
        style={{backgroundColor: '#FF0000', padding: '20px', borderRadius: '8px'}}
      />,
    );
    const button = getByTestId('btn');
    expect(button).toHaveStyle('background-color: #FF0000');
    expect(button).toHaveStyle('padding: 20px');
    expect(button).toHaveStyle('border-radius: 8px');
  });

  test('ButtonStyled accepts className prop', () => {
    const {getByTestId} = render(<S.ButtonStyled data-testid="btn" className="custom-btn" />);
    const button = getByTestId('btn');
    expect(button).toHaveClass('custom-btn');
  });
});
