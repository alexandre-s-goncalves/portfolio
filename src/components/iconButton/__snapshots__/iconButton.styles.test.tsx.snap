// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`iconButton.styles ButtonStyled should match the snapshot 1`] = `
.c0 {
  display: flex;
  align-items: center;
  justify-content: center;
  border: none;
  outline: none;
  cursor: pointer;
  transition: background-color 0.3s ease;
}

.c0:hover:not(:disabled) {
  opacity: 0.8;
}

.c0:active:not(:disabled) {
  transform: scale(0.95);
}

.c0:disabled {
  cursor: not-allowed;
  opacity: 0.5;
}

.c0:focus-visible {
  outline: 2px solid rgba(0, 0, 0, 0.3);
  outline-offset: 2px;
}

<button
  class="c0"
/>
`;
